File: 13-Graph-Algorithms/java/WordLadderII.java
/*
 * Problem: Word Ladder II
 * Source:  LeetCode 126 - https://leetcode.com/problems/word-ladder-ii/
 * 
 * Approach: BFS for distance map + DFS backtrack for all shortest paths.
 * Time Complexity:  O(n * L^2 * 26) + O(paths)
 * Space Complexity: O(n * L)
 * 
 * Tags: graph, bfs, dfs, backtracking
 */

import java.util.*;

public class WordLadderII {
    
    public List<List<String>> findLadders(String beginWord, String endWord, List<String> wordList) {
        Set<String> wordSet = new HashSet<>(wordList);
        List<List<String>> result = new ArrayList<>();
        if (!wordSet.contains(endWord)) return result;
        
        Map<String, Integer> dist = new HashMap<>();
        dist.put(beginWord, 0);
        Queue<String> queue = new LinkedList<>();
        queue.offer(beginWord);
        
        while (!queue.isEmpty()) {
            String word = queue.poll();
            char[] chars = word.toCharArray();
            for (int i = 0; i < chars.length; i++) {
                char orig = chars[i];
                for (char c = 'a'; c <= 'z'; c++) {
                    chars[i] = c;
                    String next = new String(chars);
                    if (wordSet.contains(next) && !dist.containsKey(next)) {
                        dist.put(next, dist.get(word) + 1);
                        queue.offer(next);
                    }
                }
                chars[i] = orig;
            }
        }
        
        if (!dist.containsKey(endWord)) return result;
        
        List<String> path = new ArrayList<>();
        path.add(endWord);
        dfs(endWord, beginWord, dist, path, result);
        return result;
    }
    
    void dfs(String word, String beginWord, Map<String, Integer> dist, List<String> path, List<List<String>> result) {
        if (word.equals(beginWord)) {
            List<String> reversed = new ArrayList<>(path);
            Collections.reverse(reversed);
            result.add(reversed);
            return;
        }
        
        char[] chars = word.toCharArray();
        for (int i = 0; i < chars.length; i++) {
            char orig = chars[i];
            for (char c = 'a'; c <= 'z'; c++) {
                chars[i] = c;
                String prev = new String(chars);
                if (dist.containsKey(prev) && dist.get(prev) == dist.get(word) - 1) {
                    path.add(prev);
                    dfs(prev, beginWord, dist, path, result);
                    path.remove(path.size() - 1);
                }
            }
            chars[i] = orig;
        }
    }
    
    public static void main(String[] args) {
        WordLadderII sol = new WordLadderII();
        List<String> wordList = Arrays.asList("hot","dot","dog","lot","log","cog");
        List<List<String>> result = sol.findLadders("hit", "cog", wordList);
        System.out.println("All shortest paths: " + result);
    }
}

File: 13-Graph-Algorithms/cpp/word_ladder_ii.cpp
/*
 * Problem: Word Ladder II
 * Source:  LeetCode 126 - https://leetcode.com/problems/word-ladder-ii/
 * 
 * Approach: BFS to find shortest distance from beginWord to all words.
 *           Then DFS backtrack from endWord following decreasing distances
 *           to collect all shortest paths.
 * Time Complexity:  O(n * L^2 * 26) for BFS + O(paths) for DFS
 * Space Complexity: O(n * L)
 * 
 * Tags: graph, bfs, dfs, backtracking
 */

#include <bits/stdc++.h>
using namespace std;

vector<vector<string>> findLadders(string beginWord, string endWord, vector<string>& wordList) {
    unordered_set<string> wordSet(wordList.begin(), wordList.end());
    if (!wordSet.count(endWord)) return {};
    
    // BFS to build distance map
    unordered_map<string, int> dist;
    dist[beginWord] = 0;
    queue<string> q;
    q.push(beginWord);
    
    while (!q.empty()) {
        string word = q.front(); q.pop();
        
        for (int i = 0; i < (int)word.size(); i++) {
            string next = word;
            for (char c = 'a'; c <= 'z'; c++) {
                next[i] = c;
                if (wordSet.count(next) && !dist.count(next)) {
                    dist[next] = dist[word] + 1;
                    q.push(next);
                }
            }
        }
    }
    
    if (!dist.count(endWord)) return {};
    
    // DFS backtrack from endWord
    vector<vector<string>> result;
    vector<string> path = {endWord};
    
    function<void(const string&)> dfs = [&](const string& word) {
        if (word == beginWord) {
            vector<string> reversed(path.rbegin(), path.rend());
            result.push_back(reversed);
            return;
        }
        
        for (int i = 0; i < (int)word.size(); i++) {
            string prev = word;
            for (char c = 'a'; c <= 'z'; c++) {
                prev[i] = c;
                if (dist.count(prev) && dist[prev] == dist[word] - 1) {
                    path.push_back(prev);
                    dfs(prev);
                    path.pop_back();
                }
            }
        }
    };
    
    dfs(endWord);
    return result;
}

int main() {
    vector<string> wordList = {"hot","dot","dog","lot","log","cog"};
    string beginWord = "hit", endWord = "cog";
    
    auto result = findLadders(beginWord, endWord, wordList);
    
    cout << "All shortest paths:" << endl;
    for (auto& path : result) {
        for (auto& w : path) cout << w << " -> ";
        cout << endl;
    }
    
    return 0;
}
